src/Inciter/DG.cpp

```diff
@@ -1115,6 +1091,97 @@ DG::comsol( int fromch,
+  // cells
+  for (const auto& b : m_bid) {
+    Assert( m_uc[0][b.second].size() == m_u.nprop(), "ncomp size mismatch" );
+    Assert( m_pc[0][b.second].size() == m_p.nprop(), "ncomp size mismatch" );
+    for (std::size_t c=0; c<m_u.nprop(); ++c) {
+      m_u(b.first,c,0) = m_uc[0][b.second][c];
+    }
+    for (std::size_t c=0; c<m_p.nprop(); ++c) {
+      m_p(b.first,c,0) = m_pc[0][b.second][c];
+    }
+  }
+
+  if (pref && m_stage == 0 && d->T() > 0)
+  {
+    eval_ndof( m_nunk, Disc()->Coord(), Disc()->Inpoel(), m_fd, m_u,
+               g_inputdeck.get< tag::pref, tag::indicator >(),
+               g_inputdeck.get< tag::discr, tag::ndof >(),
+               g_inputdeck.get< tag::pref, tag::ndofmax >(),
+               g_inputdeck.get< tag::pref, tag::tolref >(),
+               m_ndof );
+
+    // communicate ndof of ghost (if any)
+    if (m_ghostData.empty())
+      comdof_complete();
+    else
+      for(const auto& n : m_ghostData) {
+        std::vector< std::size_t > tetid( n.second.size() );
+        std::vector< std::size_t > ndof;
+        std::size_t j = 0;
+        for(const auto& i : n.second) {
+          Assert( i.first < m_fd.Esuel().size()/4, "Sending solution ghost data" );
+          tetid[j] = i.first;
+          ndof.push_back( m_ndof[i.first] );
+          ++j;
+        }
+        thisProxy[ n.first ].comdof( thisIndex, tetid, ndof );
+      }
+    }
+    // when this is not the last stage of RK time integration or p-daptive is
+    // not applied which means no updates to ndof data, there is no need to
+    // communicate ndof
+    else
+      comdof_complete();
+
+  owndof_complete();
+}
+
+void DG::comdof( int fromch,
+                 const std::vector< std::size_t >& tetid,
+                 const std::vector< std::size_t >& ndof )
+// *****************************************************************************
+//  Receive ndof ghost data from neighboring chares
+//! \param[in] fromch Sender chare id
+//! \param[in] tetid Ghost tet ids we receive solution data for
+//! \param[in] ndof Number of degrees of freedom for chare-boundary elements
+//! \details This function receives number of degrees of freedom from fellow
+//!   chares.
+// *****************************************************************************
+{
+  Assert( ndof.size() == tetid.size(), "Size mismatch in DG::comsol()" );
+
+  // Find local-to-ghost tet id map for sender chare
+  const auto& n = tk::cref_find( m_ghost, fromch );
+
+  for (std::size_t i=0; i<tetid.size(); ++i) {
+    auto j = tk::cref_find( n, tetid[i] );
+    Assert( j >= m_fd.Esuel().size()/4, "Receiving solution non-ghost data" );
+    auto b = tk::cref_find( m_bid, j );
+    Assert( b < m_ndofc[0].size(), "Indexing out of bounds" );
+    m_ndofc[0][b] = ndof[i];
+  }
+
+  // if we have received all ndof ghost contributions from those chares we
+  // communicate along chare-boundary faces with, solve the system
+  if (++m_nsol == m_ghostData.size()) {
+    m_nsol = 0;
```

Comment:
You are reusing the counter `m_nsol` for detecting completeness of the communication for `comdof`, and the same is used for `comsol`. These will overwrite each other if/when `comsol` and `comdof` are overlapped -- unless you make sure they cannot overlap. I would use a new counter.